# Processamento de textos com tranformers

Modelo bert minimizado com diferença entre palavras com caixa alta e baixa

In [3]:
from transformers import pipeline
pipe = pipeline(model='distilbert/distilbert-base-cased', task='feature-extraction', 
                return_tensors=True)

d:\Programs\miniconda3\envs\vc2024\Lib\site-packages\huggingface_hub\file_download.py:1150: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


config.json:   0%|          | 0.00/465 [00:00<?, ?B/s]

d:\Programs\miniconda3\envs\vc2024\Lib\site-packages\huggingface_hub\file_download.py:159: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\letic\.cache\huggingface\hub\models--distilbert--distilbert-base-cased. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


model.safetensors:   0%|          | 0.00/263M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

In [4]:
text = 'Some sentence to test the model'
y = pipe(text)

In [5]:
y[0].shape

torch.Size([8, 768])

**Tokenizador**

Codificação do texto em representação numérica

**Modelo**

Modelo baseado em encoders (bert) para realização de uma tarefa de ML

In [6]:
tokenizer = pipe.tokenizer
model = pipe.model

print(tokenizer)
print(model)

DistilBertTokenizerFast(name_or_path='distilbert/distilbert-base-cased', vocab_size=28996, model_max_length=512, is_fast=True, padding_side='right', truncation_side='right', special_tokens={'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'}, clean_up_tokenization_spaces=True),  added_tokens_decoder={
	0: AddedToken("[PAD]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	100: AddedToken("[UNK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	101: AddedToken("[CLS]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	102: AddedToken("[SEP]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	103: AddedToken("[MASK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
}
DistilBertModel(
  (embeddings): Embeddings(
    (word_embeddings): Embedding(28996, 768, padding_idx=0)
 

Em vez de mapear todas as palavras existentes, os tokens podem representar subpalavras.

Certos tokens, como '##ko' e '##herence' representam subwords, em que '##' representa que este token é uma continuação do token anterior.

In [8]:
vocab = tokenizer.vocab
vocab

{'Ridley': 23013,
 '1867': 6988,
 'Courts': 16624,
 'Computer': 6701,
 'satire': 20817,
 '##ko': 2718,
 'laughed': 3348,
 'Hawaii': 6826,
 'architects': 10939,
 '1879': 6917,
 '##rve': 24941,
 'Ah': 7066,
 'honor': 3874,
 'scouting': 26820,
 '1982': 2294,
 'mature': 9881,
 'Winnipeg': 11070,
 'knight': 11295,
 'ISO': 11533,
 'disabilities': 16292,
 '##herence': 21634,
 'Judd': 21971,
 'listed': 2345,
 'processed': 14659,
 '##por': 18876,
 '##p': 1643,
 '161': 18563,
 'tables': 7072,
 'generates': 21241,
 'contraction': 24827,
 'Langdon': 16000,
 'bush': 13771,
 '∈': 850,
 'shoving': 18164,
 '##olar': 21459,
 'expanded': 3631,
 'printed': 5757,
 'Penguins': 24801,
 '##zuki': 24849,
 'Lagos': 17883,
 'collar': 9704,
 'frightened': 10551,
 'Sister': 9536,
 'decreases': 19377,
 'sexy': 8141,
 'wishing': 10863,
 '1983': 2278,
 'soprano': 11264,
 'mandatory': 11839,
 'deficit': 16312,
 'goose': 20398,
 'processors': 20026,
 'minority': 7309,
 'Joyce': 11882,
 'stream': 5118,
 '##ec': 10294,


In [15]:
text = 'Some sentence to test the model. Você conhece o modelo?\n\nNão'
y = pipe(text)
y

tensor([[[ 0.2150,  0.1129, -0.0178,  ...,  0.0261,  0.1722,  0.1717],
         [-0.0274, -0.3160, -0.0077,  ..., -0.0538,  0.2889,  0.1436],
         [ 0.0364,  0.0265,  0.1916,  ..., -0.0422,  0.0403,  0.1278],
         ...,
         [-0.2157,  0.2550, -0.1105,  ..., -0.1735,  0.3528,  0.4155],
         [-0.0497, -0.0609,  0.0962,  ...,  0.3128, -0.0523,  0.4109],
         [-0.2265, -0.1573, -0.4688,  ..., -0.2513,  0.8715,  0.0113]]])

In [16]:
text.encode()

b'Some sentence to test the model. Voc\xc3\xaa conhece o modelo?\n\nN\xc3\xa3o'

A tokenização é realizada fazendo a codificação de cada token à um índice do vocabulário

In [17]:
tokens = tokenizer.encode(text)
tokens

[101,
 1789,
 5650,
 1106,
 2774,
 1103,
 2235,
 119,
 159,
 13335,
 24559,
 14255,
 4638,
 2093,
 184,
 2235,
 1186,
 136,
 151,
 9290,
 102]

Tokens especiais como [CLS] e [SEP] são adicionados.

0 -> [PAD]: padding para completar o tamanho da sentença

100 -> [UNK]: token que não está presente no vocabulário

101 -> [CLS]: inicio de sentença

102 -> [SEP]: separador

103 -> [MASK]: máscara usada no treinamento

In [18]:
tokenizer(text).tokens()

['[CLS]',
 'Some',
 'sentence',
 'to',
 'test',
 'the',
 'model',
 '.',
 'V',
 '##oc',
 '##ê',
 'con',
 '##he',
 '##ce',
 'o',
 'model',
 '##o',
 '?',
 'N',
 '##ão',
 '[SEP]']

In [19]:
texts = [text, 'Another sentence']
tokens = tokenizer(texts, return_tensors='pt', padding=True)
tokens

{'input_ids': tensor([[  101,  1789,  5650,  1106,  2774,  1103,  2235,   119,   159, 13335,
         24559, 14255,  4638,  2093,   184,  2235,  1186,   136,   151,  9290,
           102],
        [  101,  2543,  5650,   102,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])}

## Embeddings

Os ids dos tokens são transformados em vetores de valores. O número de linhas é o tamanho do vocabulário

In [21]:
embedding = model.embeddings(tokens['input_ids'])

embedding.shape

torch.Size([2, 21, 768])

O modelo gera duas matrizes, uma para cada sentença. Cada sentença possui 8 tokens. Para cada token, são gerados 768 atributos.

In [23]:
res = model(**tokens)
res[0].shape

torch.Size([2, 21, 768])

In [24]:
text_features = res[0][:,0]
text_features.shape

torch.Size([2, 768])

## Codificador de texto

In [22]:
from torch import nn
from transformers import pipeline

In [ ]:
class TextEncoder(nn.Module):
    
    def __init__(self):
        
        
        pipe = pipeline(model='distilbert/distilbert-base-cased', task='feature-extraction', 
                return_tensors=True)
        tokenizer = pipe.tokenizer
        model = pipe.model
        
        self.tokenizer = tokenizer
        self.model = model
        self.cls_token_id = 0
        self.feature_dim = 768
        
    def forward(self, text):
        
        tokens = self.tokenizer(text, (list, tuple))
        